app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt

```diff
@@ -0,0 +1,40 @@
+package it.unibo.sc
+
+import android.util.Log
+import androidx.paging.PagingSource
+import androidx.paging.PagingState
+import com.amplifyframework.datastore.generated.model.ProductWarehouse
+import it.unibo.sc.queries.ListProductsWarehouseQuery
+
+/**
+ * Class that loads pages of ProductWarehouse by page token.
+ */
+class ProductsWarehousePagingSource(private val graphQLData: ListProductsWarehouseQuery) :
+    PagingSource<String, ProductWarehouse>() {
+    override suspend fun load(params: LoadParams<String>): LoadResult<String, ProductWarehouse> {
+        return try {
+            val page = params.key
+            val response = graphQLData.getProductsWarehouse(page)
+            val items = response?.items?.toList() ?: emptyList()
+            val nextToken = response?.requestForNextResult?.variables?.get("nextToken")
```

Comment:
Also here: safe call but no fallback value

Reply:
the fallback value in this case would be "null" which is an accepted value (as in grapqhl a nextToken = null means last page) so elvis operator would be redundant